crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -1257,6 +1124,195 @@ impl Client {
+                                    continue 'next_peer;
+                                }
+                            }
+                        };
+
+                        if current_count == 0 {
+                            // The counter for this block has been exhausted which means
+                            // that this block is complete.
+                            tracing::trace!(block_number=%start, "All transactions received for block");
+
+                            yield PeerData::new(
+                                peer,
+                                UnverifiedTransactionData {
+                                    expected_commitment: std::mem::take(
+                                        &mut current_commitment
+                                    ),
+                                    transactions: std::mem::take(&mut transactions),
+                                    block_number: start,
+                                },
+                            );
+
+                            if start < stop {
+                                // Move to the next block
+                                start += 1;
+                                tracing::trace!(next_block=%start, "Moving to next block");
+                                let (count, commitment) = transaction_counts_and_commitments_stream
+                                .next()
+                                .await
+                                .with_context(|| {
+                                    format!(
+                                        "Transaction counts and commitments stream terminated \
+                                        prematurely at block {}",
+                                        start
+                                    )
+                                })
+                                .map_err(peer_err)?
+                                .map_err(peer_err)?;
+
+                                current_count = count;
+                                current_count_outer = Some(current_count);
+                                current_commitment = commitment;
+
+                                tracing::trace!(%start, %current_count, ?current_count_outer, "1 Getting counts from stream");
+                            }
+                        }
+                    }
+
+                    // TODO punish the peer
+                    // If we reach here, the peer did not send a Fin, so the counter for the current block should be reset
+                    // and we should start from the current block again but from the next peer
+                    //
+                    // The problem here is that the count stream was already consumed, so we assume that the full blocks that were already
+                    // processed are correct and we start from the next block from the next peer
+
+                    tracing::error!(peer=?TestPeer(peer), ?current_count, ?current_count_outer, "NO FIN FOR THIS PEER");
```

Comment:
```suggestion
                    tracing::debug!(peer=?TestPeer(peer), ?current_count, ?current_count_outer, "No Fin sent by peer");
```

I think this should not be an error. In general I think we should avoid logging an error for anything that's actually triggered by a specific network behavior of peers/clients.

We could punish the peer instead since this is clearly a protocol error.